### PR TITLE
Prometheus dashboard add-on kube application ( #3118 )

### PR DESCRIPTION
**Pull request by jayunit100**
Implementation of issue #3118.

Here's a first pass at promdash + prometheus integration into addons.

With this, you can pretty much just run both out of the box, it layers on top of the work done already by @mwielgus.  

See the README for directions, mostly its just (1) run a docker command locally and then (2) create the pod.  

I think we can make this into a replication controller as a next step.

cc @timothysc @rrati

### CODE REVIEW

**Comment by jayunit100** [comment]
For those new to prometheus and also for the sake of verification, I've attached a screenshot of a working promdash/prometheus pair which spins up from following the README directions... 

screenshot from 2015-05-06 23 50 47

**Comment by wojtek-t** [comment]
cc @fgrzadkowski FYI

**Comment by mbforbes** [comment]
shot in the dark, but assigning to @zmerlynn for addons?

**Comment by jayunit100** [comment]
Hi.  I updated the boilerplate checker, so that it only verifies license header in `.sh` files.  I put that as a *separate* commit.

**Comment by eparis** [comment]
I'd rather see the boilerplate change as a separate PR.  I'm still trying to decide how I feel about it...

File: cluster/addons/prometheus/prometheus.yaml [hunk 1]
```diff
@@ -0,0 +1,41 @@
+apiVersion: v1beta1
```
Comment (timothysc) [inline cluster/addons/prometheus/prometheus.yaml]: why not beta3 for *?

File: cluster/addons/prometheus/prometheus.yaml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    version: v1beta1
+    containers:
+      - name: kube-promdash
+        image: prom/promdash
+        ports:
+          - containerPort: 3000
+            hostPort: 3000
+        env:
+          - name: "DATABASE_URL"
+            value: "sqlite3:/promdash/file.sqlite3"
```
Comment (timothysc) [inline cluster/addons/prometheus/prometheus.yaml]: Can we comment out a block for folks to enable volume mounts for persistence. 

/cc @markturansky

File: cluster/addons/prometheus/run_prometheus.sh [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  echo "-d Prometheus' root directory (i.e. where config file/metrics data will be stored)."
+}
+
+build_config() {
+  echo >$1 'global: { scrape_interval: "10s" evaluation_interval: "10s"}'
+  local target
+
+  ### Allow prometheus to scrape multiple targets
+  ### i.e. -t KUBERNETES_RO,MY_METRICS_ABC
+  for target in ${2//,/ }; do
```
Comment (eparis) [inline cluster/addons/prometheus/run_prometheus.sh]: Is this a bash-ism, or it generic enough that leaving /bin/sh at the top is ok?

File: cluster/addons/prometheus/README.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+By default the metrics are written to a temporary location (that can be changed
+in the the volumes section of the yaml file). Prometheus' UI is available 
+at port 9090.
+
+# TODO
+
+- We should publish this image into the kube/ namespace.
+- Possibly use postgre or mysql as a promdash database.
+- Setup high availability via NFS
```
Comment (timothysc) [inline cluster/addons/prometheus/README.md]: push gateway (https://github.com/prometheus/pushgateway)?

**Comment by timothysc** [comment]
@vishh - eventually I could see this as a roll up with heapster.

**Comment by vishh** [comment]
@timothysc: Ack. I plan to start working on prometheus support in heapster soon. I just hope we don't re-invent heapster and end up maintaining one more component.

**Comment by timothysc** [comment]
@vishh +1

**Comment by jayunit100** [comment]
I'll implement these changes and update to bash and retest........ then will update pr

**Comment by jayunit100** [comment]
okay ! Implemented the v1beta3 , tested, did some minor cleanups, rebased, removed my boilerplate changes, and layered commits properly.  Ready for final  review !

**Comment by jayunit100** [comment]
hi, @zmerlynn @timothysc looks good to merge now ?  would like to run this as a pod in our cluster starting monday.

**Comment by timothysc** [comment]
@fgrzadkowski @zmerlynn @quinton-hoole ping ^

**Comment by quinton-hoole** [comment]
Sorry, we're all a bit swamped with v1.0 blockers at the moment - will take me a few days to get to this.

File: cluster/addons/prometheus/README.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
+# Prometheus in Kubernetes
+
+This is an experimental [Prometheus](http://prometheus.io/) setup for monitoring
+Kubernetes services that expose prometheus-friendly metrics through address
+http://service_address:service_port/metrics.
+
+# Purpose
+The purpose of the setup is to gather performance-related metrics during load
```
Comment (bprashanth) [inline cluster/addons/prometheus/README.md]: Its a little confusing that this is a cluster addon, for this to be useful doesn't it have to be a standalone thing that multiple load tests can send their metrics to across a week? how do other e2e clusters discover the dashboard?

**Comment by jayunit100** [comment]
the metrics are already made available by the api server and they are in prometheus format ... prom dash and prometheus just scrape them from the kubkubernetes_ro endpoints

**Comment by jayunit100** [comment]
Clarification: prometheus scrapes... prom dash gives a graphing ui

**Comment by bprashanth** [comment]
I guess im asking if the dashboard is brouht up/down with each test run? (haven't looked at your code yet, but sounds like that must be the case unless you're always scraping a known apiserver endpoint that's the same for all test clusters).

**Comment by jayunit100** [comment]
good question. so the way it works is that the kubernetes_ro endpoint is always available to any container... and that is the magic that allows the prometheus server to launch from a pod.  Makes sense?

**Comment by bprashanth** [comment]
That I know :)
'pologies, when i said test run i meant test suite. I would find a dashboard that all my e2e test suites could send, eg: latency of a list/watch to very useful (or aggregate duration of all the lists from all the tests for a whole week).

**Comment by jayunit100** [comment]
This Sounds like a specific question about a generic feature add on ... i was hoping folks would evolve this as a base starting point : there are many ways to leverage prometheus

**Comment by timothysc** [comment]
@bprashanth We currently use Prometheus as an external component to our cluster, b/c we tare-down and rebuild our cluster nightly.  By doing this, we preserve history.   

The add-on PR is meant to be a simple introspective service which "could" store it's data across runs by mounting a network volume.

**Comment by zmerlynn** [comment]
This feels more like a `contrib/` PR than

**Comment by jayunit100** [comment]
Based on experience with other projects;  Contrib/ implies it's a third party artifact which isn't necessarily tied to the core project.  Is that the case in kube?

If yes...

Prometheus is, pretty deeply integrated with kubernetes as of today... and so this is just the next logical step in supporting prometheus.

If no....

Then I guess it might belong in contrib/... but maybe then the use of contrib is for a different purpose then I've guessed above.

Either way is fine with me but I'd prefer to iterate tThis code in a place which implies it is a  supported compoment.

**Comment by zmerlynn** [comment]
Right now, everything in the `addons` directory is a pod of some sort that gets loaded by Salt through one means or another. This doesn't seem to fit that that pattern. Hm. I'm not sure where it belongs.

**Comment by timothysc** [comment]
@vishh Given that you would be the heapster integration person, do you have opinions on the matter?

**Comment by zmerlynn** [comment]
Note that what I said is kind of an artifact of just what's already been built. This is definitely an addon, but everything else in that dir so far is something we have an option of launching at cluster creation. I suppose this isn't too far off the first steps of Elasticsearch, though.

**Comment by vishh** [comment]
IIUC, Kubernetes is not deeply tied to Prometheus. The fact that we use
prometheus client libraries does not mean we are tied to Prometheus. We
could very well move to a different metrics client library in the future.
This PR feels like a Contrib/ to me as well. We are trying to bring down
the number of external dependencies in the project and another addon is not
ideal at this stage. The core system can function without a monitoring
backend. But the existing addons like kube2sky and heapster are required
for the core to function.

Once heapster scrapes the component metrics, we can use other monitoring
solutions too for this purpose.

**Comment by jayunit100** [comment]
ok.  well if the consensus to move it to `contrib/` i can do that.  anyone else want to vote ? looks like 2 in favor of contrib, 1 in favor of addon.

**Comment by jayunit100** [comment]
ok, moved to contrib & rebased :)

**Comment by timothysc** [comment]
thx jay! 

@zmerlynn any other comments?

**Comment by keegancsmith** [comment]
This was merged in a while ago, but not sure of a better place to bring this up. Prometheus now has built in Kubernetes discovery and the example config will automatically scrape all the `kube-system` services + services which have the annotation `"prometheus.io/scrape": "true"`. https://github.com/prometheus/prometheus/blob/master/documentation/examples/prometheus-kubernetes.yml

**Comment by jayunit100** [comment]
Hi keegan thanks for bringing this up.  We have prometheus and promdash
integrated in the contrib/prometheus stuff. So long as the functionality is
retained, I think replacing parts with the idioms that re being used in
github.com/prometheus/... makes sense.  As I recall, the prometheus docker
images were out of data and promdash was a 3rd party extension at the time
that we created this.




On Thu, Oct 1, 2015 at 5:36 AM, Keegan Carruthers-Smith <
notifications@github.com> wrote:

> This was merged in a while ago, but not sure of a better place to bring
> this up. Prometheus now has built in Kubernetes discovery and the example
> config will automatically scrape all the kube-system services + services
> which have the annotation "prometheus.io/scrape": "true".
> https://github.com/prometheus/prometheus/blob/master/documentation/examples/prometheus-kubernetes.yml
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/7878#issuecomment-144670656>
> .
>



-- 
jay vyas

**Comment by fgrzadkowski** [comment]
FYI @mwielgus @jszczepkowski @piosz

**Pull request merged**